gradle/shared-with-buildSrc/mirrors.gradle.kts

```diff
@@ -14,34 +14,71 @@
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
 
-val repositoryMirrors = findMirrorUrls()
-rootProject.extensions.extraProperties.set("repositoryMirrors", repositoryMirrors)
 
-val gradleMirrorUrl = repositoryMirrors["gradle"]
-if (gradleMirrorUrl != null) {
-    project.allprojects {
-        buildscript.repositories {
-            maven {
-                name = "gradle-mirror"
-                url = uri(gradleMirrorUrl)
-            }
-        }
-        repositories {
-            maven {
-                name = "gradle-mirror"
-                url = uri(gradleMirrorUrl)
+val originalUrls: Map<String, String> = mapOf(
+    "jcenter" to "https://jcenter.bintray.com/",
+    "mavencentral" to "https://repo.maven.apache.org/maven2/",
+    "google" to "https://dl.google.com/dl/android/maven2/",
+    "gradle" to "https://repo.gradle.org/gradle/repo",
+    "gradleplugins" to "https://plugins.gradle.org/m2",
+    "gradlejavascript" to "https://repo.gradle.org/gradle/javascript-public",
+    "gradle-libs" to "https://repo.gradle.org/gradle/libs",
+    "gradle-releases" to "https://repo.gradle.org/gradle/libs-releases",
+    "gradle-snapshots" to "https://repo.gradle.org/gradle/libs-snapshots",
+    "kotlindev" to "https://dl.bintray.com/kotlin/kotlin-dev/",
+    "kotlineap" to "https://dl.bintray.com/kotlin/kotlin-eap/"
+)
+
+val mirrorUrls: Map<String, String> =
+    System.getenv("REPO_MIRROR_URLS")?.split(',')?.associate { nameToUrl ->
+        val (name, url) = nameToUrl.split(':', limit = 2)
+        name to url
+    } ?: emptyMap()
+
+fun withMirrors(handler: RepositoryHandler) {
+    if ("CI" !in System.getenv()) {
+        return
+    }
+    handler.all {
+        if (this is MavenArtifactRepository) {
+            originalUrls.forEach { name, orignalUrl ->
+                if (normalizeUrl(orignalUrl) == normalizeUrl(this.url.toString()) && mirrorUrls.containsKey(name)) {
+                    this.setUrl(mirrorUrls.get(name))
+                }
             }
         }
     }
 }
 
-fun findMirrorUrls(): Map<String, String> =
-    if ("CI" in java.lang.System.getenv()) {
-        System.getenv("REPO_MIRROR_URLS")?.split(',')?.associate { nameToUrl ->
-            val (name, url) = nameToUrl.split(':', limit = 2)
-            name to url
-        } ?: emptyMap()
-    } else {
-        emptyMap()
+fun normalizeUrl(url: String): String {
+    val result = url.replace("https://", "http://")
+    return if (result.endsWith("/")) result else result + "/"
+}
+
+project.allprojects {
+    buildscript.configurations["classpath"].incoming.beforeResolve {
+        withMirrors(buildscript.repositories)
     }
+    afterEvaluate {
+        withMirrors(repositories)
+    }
+}
+
+val settings = (gradle as org.gradle.api.internal.GradleInternal).settings
+withMirrors(settings.pluginManagement.repositories)
```

Comment:
❌ I think that is too late. We should apply the script to `settings.gradle.kts` and then access `Gradle` as we do in the integration test mirror init script. To get IDE support, I think the name of the script needs to be `mirrors.settings.gradle.kts`.

Reply:
I don't think so. I did an experiment on a brand new cloud server with everything new, then change the repos we used in `/etc/hosts` to `127.0.0.1`, then run the build, everything works. I didn't dive into the details but I think this did work.

Reply:
The advantage of applying the script to `settings.gradle.kts` is that you don't need to resort to using internal APIs. I guess using `settingsEvaluated` from `build.gradle.kts` has no effect, right?

Why shouldn't we apply this script to `settings.gradle.kts` instead of `build.gradle.kts`? For me setting up the mirrors for the whole build belongs more to `settings.gradle.kts` than `build.gradle.kts`.